### PR TITLE
Bpp and redis annotation

### PR DESCRIPTION
**Pull request by afanasievM**
added beanPostProcessor that runs request to get currency on startup.
Added CustomCacheable that cached currency rates for 1 min

### CODE REVIEW

**Comment by afanasievM** [comment]
@ajax-gnatiuk-s

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/annotation/CustomCacheable.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package ua.com.foxminded.restClient.annotation
+
+@Target(AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.RUNTIME)
+annotation class CustomCacheable(
+    val value: String
+)
+
+
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/annotation/CustomCacheable.kt]: one line break at the end of file is enough :)

File: university-service/src/test/kotlin/ua/com/foxminded/courseproject/validation/ValidationSetupTest.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package ua.com.foxminded.courseproject.validation
+
+import javax.validation.Validation
+import javax.validation.Validator
+
+open class ValidationSetupTest {
+
+    protected var validator: Validator = Validation.buildDefaultValidatorFactory().validator
+
+}
```
Comment (ajax-gnatiuk-s) [inline university-service/src/test/kotlin/ua/com/foxminded/courseproject/validation/ValidationSetupTest.kt]: looks like incomplete test, isn't it?

File: university-service/build.gradle.kts [hunk 3]
```diff
@@ -74,4 +83,12 @@ tasks.withType<Javadoc>() {
...
     jvmToolchain(17)
+}
+
+tasks.withType<Detekt>().configureEach {
+    jvmTarget = "17"
+}
+
+tasks.withType<DetektCreateBaselineTask>().configureEach {
+    jvmTarget = "17"
 }
```
Comment (ajax-gnatiuk-s) [inline university-service/build.gradle.kts]: no line break at the and of file

File: university-service/build.gradle.kts [hunk 4]
```diff
@@ -1,3 +1,7 @@
+import io.gitlab.arturbosch.detekt.DetektCreateBaselineTask
+import io.gitlab.arturbosch.detekt.Detekt
+
+
 /*
  * This file was generated by the Gradle 'init' task.
```
Comment (ajax-gnatiuk-s) [inline university-service/build.gradle.kts]: I don't think we really need this comment

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/WebConfig.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.web.reactive.config.EnableWebFlux
+import org.springframework.web.reactive.config.WebFluxConfigurer
+import org.springframework.web.reactive.result.method.annotation.ArgumentResolverConfigurer
+
+
+@Configuration
+@EnableWebFlux
+class WebConfig: WebFluxConfigurer {
+    @Value("\${nats.url}")
+    private lateinit var natsUrl: String
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/WebConfig.kt]: why not passing the property to `#natsConncetion`?

```kotlin
    @Bean
    fun natsConncetion(@Value("\${nats.url}") natsUrl: String): Connection {
        return Nats.connect(natsUrl)
    }
```

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/RedisConfig.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                                mapper
+                                    .copy()
+                                    .enableDefaultTyping(
+                                        ObjectMapper.DefaultTyping.NON_FINAL, JsonTypeInfo.As.PROPERTY
+                                    )
+                            )
+                        )
+                    )
+            )
+            .build()
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/RedisConfig.kt:37]: it's hard to read the code.

I suggest to extract few variables, for example:
```kotlin
val jackson2JsonRedisSerializer = GenericJackson2JsonRedisSerializer(...)
val redisSerializationContext = RedisSerializationContext.SerializationPair.fromSerializer(
    jackson2JsonRedisSerializer
)
```

File: university-service/build.gradle.kts (hunk 3 above)
Comment (afanasievM) [reply university-service/build.gradle.kts]: ✅

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/RedisConfig.kt (hunk 6 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/RedisConfig.kt:37]: ✅

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/WebConfig.kt (hunk 5 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/ua/com/foxminded/restClient/config/WebConfig.kt]: Interesting. I did'n see it before. 
Changed.

File: university-service/build.gradle.kts (hunk 4 above)
Comment (afanasievM) [reply university-service/build.gradle.kts]: ✅

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/annotation/CustomCacheable.kt (hunk 1 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/ua/com/foxminded/restClient/annotation/CustomCacheable.kt]: ✅

File: university-service/src/test/kotlin/ua/com/foxminded/courseproject/validation/ValidationSetupTest.kt (hunk 2 above)
Comment (afanasievM) [reply university-service/src/test/kotlin/ua/com/foxminded/courseproject/validation/ValidationSetupTest.kt]: May be "ValidationTestConfig"? What do you think? 
It isn't test class. It is just presetup. Or may be add line 
protected var validator: Validator = Validation.buildDefaultValidatorFactory().validator
to all validation tests.

Comment (afanasievM) [reply university-service/src/test/kotlin/ua/com/foxminded/courseproject/validation/ValidationSetupTest.kt]: chanched to testfixtures

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**